test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests/SignatureTrustAndValidityVerificationProviderTests.cs

```diff
@@ -1521,6 +1521,263 @@ public async Task GetTrustResultAsync_WithExpiredRepositoryCertificateAndNoTimes
             }
         }
 
+        [Collection(SigningTestCollection.Name)]
+        public class FallbackFromPrimarySignaturesToCountersignatures
+        {
+            //The settings when validating packages from nuget.org in accept mode (AcceptModeDefaultPolicy + allowUnsigned:false + allowUntrusted:false)
+            private readonly SignedPackageVerifierSettings _defaultNuGetOrgSettings = new SignedPackageVerifierSettings(
+                allowUnsigned: false,
+                allowIllegal: true,
+                allowUntrusted: false,
+                allowIgnoreTimestamp: true,
+                allowMultipleTimestamps: true,
+                allowNoTimestamp: true,
+                allowUnknownRevocation: true,
+                reportUnknownRevocation: false,
+                verificationTarget: VerificationTarget.All,
+                signaturePlacement: SignaturePlacement.Any,
+                repositoryCountersignatureVerificationBehavior: SignatureVerificationBehavior.IfExistsAndIsNecessary,
+                revocationMode: RevocationMode.Online);
+            private readonly SigningTestFixture _fixture;
+            private readonly SignatureTrustAndValidityVerificationProvider _provider;
+
+            public FallbackFromPrimarySignaturesToCountersignatures(SigningTestFixture fixture)
+            {
+                if (fixture == null)
+                {
+                    throw new ArgumentNullException(nameof(fixture));
+                }
+
+                _fixture = fixture;
+                _provider = new SignatureTrustAndValidityVerificationProvider();
+            }
+
+            // Case1: primary signature (trusted + non-expired) doesn't fall back to countersiganture (trusted + non-expired).
+            // The verification result is the primary signature status(valid). 
+            [CIOnlyFact]
+            public async Task GetTrustResultAsync_FallbackFromGoodPrimarySigntuareToGoodCountersignature_NoFallbackAndReturnsValidAsync()
+            {
+                var settings = _defaultNuGetOrgSettings;
+
+                using (Test test = await Test.CreateAuthorSignedRepositoryCountersignedPackageAsync(
+                    authorCertificate: _fixture.TrustedTestCertificate.Source.Cert,
+                    repositoryCertificate: _fixture.TrustedRepositoryCertificate.Source.Cert))
+                using (var packageReader = new PackageArchiveReader(test.PackageFile.FullName))
+                {
+                    PrimarySignature primarySignature = await packageReader.GetPrimarySignatureAsync(CancellationToken.None);
+                    PackageVerificationResult status = await _provider.GetTrustResultAsync(packageReader, primarySignature, settings, CancellationToken.None);
+
+                    Assert.Equal(SignatureVerificationStatus.Valid, status.Trust);
```

Comment:
Also verify that NU3018 and NU3037 aren't in log output.

Reply:
Thanks! Added.